### PR TITLE
Wire up requests

### PR DESCRIPTION
**Pull request by gestchild**
## Who is this for?

People who want to be able to request items 

## What is it doing for them?

This builds on the UI work from #6872, replacing mock data with requests to the real api

Adds requesting functionality and puts it behind a single toggle with the login

N.B. there are a couple of TODOs around error logging and messaging to the user, but this is in a place where we can put it in front of people

### CODE REVIEW

File: catalogue/webapp/components/ConfirmItemRequest/ConfirmItemRequest.tsx [hunk 1]
```diff
@@ -66,11 +72,16 @@ const RequestDialog: FunctionComponent<RequestDialogProps> = ({
...
   confirmRequest,
   setIsActive,
+  currentHoldNumber,
 }) => (
   <Request isLoading={isLoading}>
     <Header>
       <span className={`h2`}>Request item</span>
-      <Remaining>7/15 items remaining</Remaining>
+      <Remaining>
+        {`${
```
Comment (jamesgorrie) [inline catalogue/webapp/components/ConfirmItemRequest/ConfirmItemRequest.tsx:81]: worth putting into something shared?

File: catalogue/webapp/components/ConfirmItemRequest/ConfirmItemRequest.tsx [hunk 2]
```diff
@@ -176,6 +193,7 @@ const ErrorDialog: FunctionComponent<ErrorDialogProps> = ({ setIsActive }) => (
       <span className={`h2`}>Request failed</span>
     </Header>
     <p className="no-margin">
+      {/* TODO: get error code and construct appropriate message from response */}
```
Comment (jamesgorrie) [inline catalogue/webapp/components/ConfirmItemRequest/ConfirmItemRequest.tsx:196]: Is it worth creating GH issues that this references?

File: catalogue/webapp/components/ConfirmItemRequest/ConfirmItemRequest.tsx [hunk 3]
```diff
@@ -184,37 +202,91 @@ const ErrorDialog: FunctionComponent<ErrorDialogProps> = ({ setIsActive }) => (
...
+type RequestingState = null | 'requesting' | 'confirmed' | 'error';
+
 const ConfirmItemRequest: FunctionComponent<Props> = props => {
   const openButtonRef = useRef<HTMLButtonElement>(null);
-  const { item, work, setIsActive, ...modalProps } = props;
-  const [isConfirmed, setIsConfirmed] = useState(false);
-  const [isLoading, setIsLoading] = useState(false);
-  const [isError, setIsError] = useState(false); // TODO: implement this if something goes wrong with API call
-
+  const { item, work, setIsActive, user, initialHoldNumber, ...modalProps } =
```
Comment (jamesgorrie) [inline catalogue/webapp/components/ConfirmItemRequest/ConfirmItemRequest.tsx:209]: Could we deconstruct this in the props like we do elsewhere?

File: catalogue/webapp/components/ConfirmItemRequest/ConfirmItemRequest.tsx [hunk 4]
```diff
@@ -184,37 +202,91 @@ const ErrorDialog: FunctionComponent<ErrorDialogProps> = ({ setIsActive }) => (
...
+    props;
+  const [requestingState, setRequestingState] = useState<RequestingState>(null);
```
Comment (jamesgorrie) [inline catalogue/webapp/components/ConfirmItemRequest/ConfirmItemRequest.tsx:211]: Why `null`?

File: catalogue/webapp/components/ConfirmItemRequest/ConfirmItemRequest.tsx [hunk 5]
```diff
@@ -184,37 +202,91 @@ const ErrorDialog: FunctionComponent<ErrorDialogProps> = ({ setIsActive }) => (
...
+  const [currentHoldNumber, setCurrentHoldNumber] = useState(initialHoldNumber);
   function innerSetIsActive(value: boolean) {
     if (value) {
       setIsActive(true);
-    } else if (isLoading) {
+    } else if (requestingState === 'requesting') {
       // disable close dialog button during api call
```
Comment (jamesgorrie) [inline catalogue/webapp/components/ConfirmItemRequest/ConfirmItemRequest.tsx:217]: Is this a TODO?

File: catalogue/webapp/components/ConfirmItemRequest/ConfirmItemRequest.tsx [hunk 6]
```diff
@@ -184,37 +202,91 @@ const ErrorDialog: FunctionComponent<ErrorDialogProps> = ({ setIsActive }) => (
...
-      setIsError(false);
     }
   }
 
-  function confirmRequest() {
-    setIsLoading(true);
-    // setIsError(true); // uncomment to test the error UI
+  useEffect(() => {
+    setCurrentHoldNumber(initialHoldNumber);
+  }, [initialHoldNumber]);
```
Comment (jamesgorrie) [inline catalogue/webapp/components/ConfirmItemRequest/ConfirmItemRequest.tsx:225]: Does this ever change?

File: catalogue/webapp/components/ConfirmItemRequest/ConfirmItemRequest.tsx [hunk 7]
```diff
@@ -184,37 +202,91 @@ const ErrorDialog: FunctionComponent<ErrorDialogProps> = ({ setIsActive }) => (
...
+          },
+        }
+      );
+      if (!response.ok) {
+        setRequestingState('error');
+        // TODO: something to Sentry?
+      } else {
+        setRequestingState('confirmed');
+        setCurrentHoldNumber(currentHoldNumber + 1);
+        // Getting the users current holds, following a successful request, in order to update this number doesn't work because the response we get isn't yet up to date, so changing it manually
```
Comment (jamesgorrie) [inline catalogue/webapp/components/ConfirmItemRequest/ConfirmItemRequest.tsx:251]: Why doesn't this work?

**Review by jamesgorrie — COMMENTED** [review]
A few code questions.

The PR became quite hard to follow as it's quite large and there were no tests or anything in the description to point out which part of the requesting items this is part of?

Do you think tests would make this easier to fathom all the different states etc?

File: catalogue/webapp/components/ConfirmItemRequest/ConfirmItemRequest.tsx (hunk 4 above)
Comment (gestchild) [reply catalogue/webapp/components/ConfirmItemRequest/ConfirmItemRequest.tsx:211]: just being explicit that there is no value, as it this point no request has been attempted

Comment (jamesgorrie) [reply catalogue/webapp/components/ConfirmItemRequest/ConfirmItemRequest.tsx:211]: This is inferred if you leave the type empty:

File: catalogue/webapp/components/ConfirmItemRequest/ConfirmItemRequest.tsx (hunk 5 above)
Comment (gestchild) [reply catalogue/webapp/components/ConfirmItemRequest/ConfirmItemRequest.tsx:217]: It's a comment David left explaining what's happening, but it's not that clear - I'll rewrite this function

File: catalogue/webapp/components/ConfirmItemRequest/ConfirmItemRequest.tsx (hunk 7 above)
Comment (gestchild) [reply catalogue/webapp/components/ConfirmItemRequest/ConfirmItemRequest.tsx:251]: The api response doesn't updated quickly enough, once a request is placed

File: catalogue/webapp/components/ConfirmItemRequest/ConfirmItemRequest.tsx (hunk 6 above)
Comment (gestchild) [reply catalogue/webapp/components/ConfirmItemRequest/ConfirmItemRequest.tsx:239]: yes, when the PhysicalItemDetails component renders and the userHolds are updated

Comment (jamesgorrie) [reply catalogue/webapp/components/ConfirmItemRequest/ConfirmItemRequest.tsx:239]: Worth noting (sorry for this broken record, but if we come back to it or someone else works on this, it would be helpful to know as it's not immediately obvious).

File: catalogue/webapp/components/ConfirmItemRequest/ConfirmItemRequest.tsx (hunk 7 above)
Comment (jamesgorrie) [reply catalogue/webapp/components/ConfirmItemRequest/ConfirmItemRequest.tsx:265]: Comment?

File: catalogue/webapp/components/WorkDetails/WorkDetails.tsx [hunk 8]
```diff
@@ -231,11 +247,11 @@ const WorkDetails: FunctionComponent<Props> = ({ work }: Props) => {
   const holdings = getHoldings(work);
 
   type WhereToFindItProps = {
-    showLogin: boolean;
+    enableRequesting: boolean;
   };
-  const WhereToFindIt = ({ showLogin }: WhereToFindItProps) => (
+  const WhereToFindIt = ({ enableRequesting }: WhereToFindItProps) => (
```
Comment (jamesgorrie) [inline catalogue/webapp/components/WorkDetails/WorkDetails.tsx:252]: You could just use the context here rather than prop passing

File: common/views/pages/_app.tsx [hunk 9]
```diff
@@ -294,23 +294,25 @@ const WecoApp: FunctionComponent<WecoAppProps> = ({
...
-            {pageProps.err && (
-              <ErrorPage
-                statusCode={pageProps.err.statusCode}
-                title={pageProps.err.message}
-                globalContextData={globalContextData}
-              />
-            )}
-          </OutboundLinkTracker>
-        </ThemeProvider>
+        <UserInfoProvider>
```
[the remark this answers is not in the document]
Comment (gestchild) [reply common/views/pages/_app.tsx]: Have move this from wrapping the entire app, to just where we need it

File: catalogue/webapp/components/ConfirmItemRequest/ConfirmItemRequest.tsx [hunk 10]
```diff
@@ -184,37 +211,97 @@ const ErrorDialog: FunctionComponent<ErrorDialogProps> = ({ setIsActive }) => (
   </>
 );
 
-const ConfirmItemRequest: FunctionComponent<Props> = props => {
-  const openButtonRef = useRef<HTMLButtonElement>(null);
-  const { item, work, setIsActive, ...modalProps } = props;
-  const [isConfirmed, setIsConfirmed] = useState(false);
-  const [isLoading, setIsLoading] = useState(false);
-  const [isError, setIsError] = useState(false); // TODO: implement this if something goes wrong with API call
+type RequestingState = undefined | 'requesting' | 'confirmed' | 'error';
```
Comment (jamesgorrie) [inline catalogue/webapp/components/ConfirmItemRequest/ConfirmItemRequest.tsx:214]: Would we normally have:
```suggestion
type RequestingState = 'requesting' | 'confirmed' | 'error';
```

And then reference it as `prop: RequestingState | undefined`?

**Review by jamesgorrie — APPROVED** [review]
This looks OK to me.

There's quite a lot addressed in 1 PR.

I would suggest breaking it into smaller pieces addressing points individually, especially as the commits group quite a lot of functionality too.

I know this is more pronounced as I am not very close to the work - but this does look as though we could have split the work up, given descriptions of what the work is, and potentially given screenshots or similar for clarity.

I also understand that @davidpmccormick and yourself have been working on it, which would have reduced the need for this.
